quinn/src/streams.rs

```diff
@@ -353,9 +353,11 @@ where
         buf: &mut ReadBuf<'_>,
     ) -> Poll<Result<(), ReadError>> {
         self.poll_read_generic(cx, |conn, stream| {
-            conn.inner
-                .read(stream, buf.remaining(), true)
-                .map(|val| val.map(|chunk| buf.put_slice(&chunk.bytes)))
+            Ok(conn
+                .inner
+                .read(stream, true)?
+                .next(buf.remaining())
+                .map(|chunk| buf.put_slice(&chunk.bytes)))
```

Comment:
To prevent loss of specific errors, we must check the status, stash any terminal states locally, and generate a suitable error without calling `inner.read` on the next poll. Similarly below.